src/libmongoc/src/mongoc/mongoc-read-concern.h

```diff
@@ -46,6 +46,16 @@ mongoc_read_concern_get_level (const mongoc_read_concern_t *read_concern);
 MONGOC_EXPORT (bool)
 mongoc_read_concern_set_level (mongoc_read_concern_t *read_concern,
                                const char *level);
+MONGOC_EXPORT (uint32_t)
+mongoc_read_concern_get_at_cluster_time_timestamp (
+   const mongoc_read_concern_t *read_concern);
+MONGOC_EXPORT (uint32_t)
+mongoc_read_concern_get_at_cluster_time_increment (
+   const mongoc_read_concern_t *read_concern);
+MONGOC_EXPORT (bool)
+mongoc_read_concern_set_at_cluster_time (mongoc_read_concern_t *read_concern,
+                                         uint32_t t,
+                                         uint32_t i);
```

Comment:
I made this API for setting and getting `at_cluster_time` public. Users _could_ feasibly send a read concern with level "snapshot" and "atClusterTime" specified to read from a particular point in time outside of a session. If this remains public, we'll have to document how this works and point to relevant docs in CDRIVER-4036.

Reply:
To play devil's advocate, I do not think this is a requirement of the spec and I do not see a compelling use case for the new API.

Snapshot reads setting `atClusterTime` in the read concern seems analogous to how causal consistency sets [`afterClusterTime` in the read concern](https://github.com/mongodb/specifications/blob/master/source/causal-consistency/causal-consistency.rst#server-commands).

There is no API to manually configure an `afterClusterTime` on a read concern. It is appended in the private [_mongoc_client_session_append_read_concern](https://github.com/mongodb/mongo-c-driver/blob/dad33550aa098229d395d2cf7c02f10a9f3f40c4/src/libmongoc/src/mongoc/mongoc-client-session.c#L1491:L1491).

The current draft of https://github.com/mongodb/specifications/pull/1022/files#diff-4a8e136fa35d82e76e2e53dd27491b4b06c1510164ade6ae65f545b6abf78709 does not indicate that new read concern API is needed to set `afterClusterTime`.

I do not think users can easily obtain the value of `atClusterTime` in `find` / `aggregate`. Users could use command monitoring or use the generic command helpers to parse the server reply. But we are not exposing a way to retrieve `atClusterTime` returned on a cursor. I think the expectation is that users will not be setting `atClusterTime` manually.

If there is not a compelling use case I would be inclined to add the API. But otherwise, I vote we leave it private.

Reply:
Additionally, if `atClusterTime` is handled the same way as `afterClusterTime`, I think that may simplify the implementation.

There would no longer be a need to store it on a read concern at all.

We would not need to conditionally append a read concern in a cursor. The call to `_mongoc_client_session_append_read_concern` could append the read concern if the command name is `aggregate` / `find` / `distinct`. That may also be more correct, in that it would cover the case that a user is using a generic command helper (like `mongoc_client_command_with_opts`) with `aggregate` / `find` / `distinct`.

Reply:
Thanks for the in-depth response here. Given the way `afterClusterTime` is currently added, and the simplicity + accuracy of just adding `snapshot` read concern in `_mongoc_client_session_append_read_concern`, I've decided to do just that and remove all new API on `mongoc_read_concern`. This really simplifies the whole PR 😆  

If @alcaeus still feels there's a need for users appending `atClusterTime` independently of the driver's sessions API, I can add the API back in (but I don't think it's needed to comply with the current spec).

Reply:
No, I think there's no need to add that 👍